dhis-2/dhis-web/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/security/JwtBearerTokenTest.java

```diff
@@ -0,0 +1,283 @@
+    private static NimbusJwtDecoder jwtDecoder;
+
+    @BeforeClass
+    public static void setUpClass()
+        throws JOSEException
+    {
+        DhisWebApiWebSecurityConfig.setApiContextPath( "" );
+
+        jwkList = new ArrayList<>();
+        jwkList.add( defaultRSA );
+
+        JWKSource<SecurityContext> jwkSource = ( jwkSelector, securityContext ) -> jwkSelector
+            .select( new JWKSet( jwkList ) );
+        jwsEncoder = new JwtUtils( jwkSource );
+
+        jwtDecoder = NimbusJwtDecoder.withPublicKey( defaultRSA.toRSAPublicKey() ).build();
+    }
+
+    @Before
+    public void setUp()
+    {
+        dhisJwtAuthenticationManagerResolver.setJwtDecoder( jwtDecoder );
+
+        dhisOidcProviderRepository.clear();
+    }
+
+    private Jwt createJwt( String provider, String clientId, String mappingKey, String mappingValue )
+    {
+        JoseHeader joseHeader = TestJoseHeaders.joseHeader( provider ).build();
+        JwtClaimsSet jwtClaimsSet = TestJwtClaimsSets.jwtClaimsSet( provider, clientId, mappingKey,
+            mappingValue ).build();
+
+        return jwsEncoder.encode( joseHeader, jwtClaimsSet );
+    }
+
+    @Test
+    public void testJwkEncodeEndDecode()
+        throws JOSEException
+    {
+        Jwt encodedJws = createJwt( TEST_PROVIDER_ONE_URI, CLIENT_ID_1, DEFAULT_MAPPING_CLAIM, DEFAULT_EMAIL );
+
+        assertEquals( "JWT", encodedJws.getHeaders().get( JoseHeaderNames.TYP ) );
+        assertEquals( defaultRSA.getKeyID(), encodedJws.getHeaders().get( JoseHeaderNames.KID ) );
+        assertNotNull( encodedJws.getId() );
+
+        String tokenValue = encodedJws.getTokenValue();
+
+        jwtDecoder.decode( tokenValue );
+    }
+
+    @Test
+    public void testSuccessfulRequest()
+    {
+        setupTestingProvider( CLIENT_ID_1, TEST_PROVIDER_ONE_NAME, TEST_PROVIDER_ONE_URI );
+
+        User openIDUser = createOpenIDUser( "openiduser", "openiduser@oidc.org" );
+
+        String tokenValue = createJwt( TEST_PROVIDER_ONE_URI, CLIENT_ID_1, "email", "openiduser@oidc.org" )
+            .getTokenValue();
+
+        JsonUser as = GET( tokenValue, "/me?fields=settings,id" ).content().as( JsonUser.class );
+        assertEquals( openIDUser.getUid(), as.getId() );
+    }
+
+    @Test
+    public void testMalformedToken()
+    {
+        JsonError error = GET( "NOT_A_JWT_TOKEN", "/me" ).error();
+        assertEquals( HttpStatus.UNAUTHORIZED.value(), error.getHttpStatusCode() );
+        assertEquals( "invalid_token", error.getMessage() );
+        assertEquals( "Invalid JWT serialization: Missing dot delimiter(s)",
+            error.getDevMessage() );
+    }
+
+    @Test
+    public void testExpiredToken()
+    {
+        dhisJwtAuthenticationManagerResolver.setJwtDecoder( null );
+
+        setupGoogleProvider( CORRECT_AUDIENCE_AKA_CLIENTID );
+
+        JsonError error = GET( EXPIRED_GOOGLE_JWT_TOKEN, "/me" ).error();
+
+        assertEquals( HttpStatus.UNAUTHORIZED.value(), error.getHttpStatusCode() );
```

Comment:
FYI: You can shorten these into 
```
JsonError error = GET( EXPIRED_GOOGLE_JWT_TOKEN, "/me" ).error(HttpStatus.UNAUTHORIZED);
```
which is not 100% the same - this tests the status code of the actual response while the `assertEquals` checks it in the error message - but I guess we are pretty sure we are consistent there.
```